### PR TITLE
Merge dependencies

### PR DESCRIPTION
**Pull request by rossberg**
The proposal currently depends on typed function references and typeful references in general. In addition, various examples require tail calls. This PR hence merges in both proposal repos as dependencies, as a prerequisite for merging the interpreter prototype in a follow-up.

### CODE REVIEW

**Review by slindley — APPROVED** [review]
LGTM

**Comment by fgmccabe** [comment]
I am assuming that this PR does not include the hybrid iso-recursive type proposal. 
It seems pretty clumsy to merge entire repos just to get the dependencies.

**Comment by rossberg** [comment]
Right, recursive types are currently with the GC proposal, which is not needed here.

Including the dependencies has been our standard practice for proposal repos. Once a dependency is merged into the upstream spec, a sync will make the dependency obsolete.

File: interpreter/syntax/types.ml [hunk 1]
```diff
@@ -1,11 +1,23 @@
...
+and var = SynVar of syn_var | SemVar of sem_var
+
+and nullability = NonNullable | Nullable
+and num_type = I32Type | I64Type | F32Type | F64Type
+and vec_type = V128Type
+and ref_type = nullability * heap_type
+and heap_type =
+  FuncHeapType | ExternHeapType | DefHeapType of var | BotHeapType
+and value_type =
+  NumType of num_type | VecType of vec_type | RefType of ref_type | BotType
```
Comment (fgmccabe) [inline interpreter/syntax/types.ml:16]: What is a BotType?

File: proposals/function-references/Overview.md [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+
+* Add an instruction `ref.as_non_null` that converts a nullable reference to a non-nullable one or traps if null
+
+* Add an instruction `br_on_null` that converts a nullable reference to a non-nullable one or branches if null, as well as the inverted `br_on_non_null`
+
+* Add an instruction `call_ref` for calling a function through a `ref $t`
+
+* Refine the instruction `ref.func $f` to return a typed function reference
+
+* Optionally add an instruction `func.bind` to create a closure
```
Comment (fgmccabe) [inline proposals/function-references/Overview.md:45]: I was under the impression that func.bind was not happening.

**Review by fgmccabe — APPROVED** [review]
I am ok (somewhat reluctantly TBH) to let this go through.
Some (most) of my comments have no real bearing on stack switching so there is not much point in elaborating them here.

File: interpreter/syntax/types.ml (hunk 1 above)
Comment (rossberg) [reply interpreter/syntax/types.ml:16]: The bottom type. It only occurs internally.

File: proposals/function-references/Overview.md (hunk 2 above)
Comment (rossberg) [reply proposals/function-references/Overview.md:45]: Yes, but I yet have to update the overview.

**Review by rossberg — COMMENTED** [review]
Thanks.

**Pull request merged**